### PR TITLE
Create volume from snapshot

### PR DESCRIPTION
**Pull request by lixiaoy1**
**What this PR does / why we need it**:

**BP Create_Volume_From_Snapshot.

**Special notes for your reviewer**:
This is to implement volume creation from snapshots.

**Release note**:

```release-note
Users can create volumes from snapshots.
```

### CODE REVIEW

**Comment by lixiaoy1** [comment]
@leonwanghui Rebased on development branch.

**Comment by leonwanghui** [comment]
It seems that the unit test is not right, please take a look

File: contrib/drivers/lvm/lvm.go [hunk 1]
```diff
@@ -110,6 +112,24 @@ func (d *Driver) CreateVolume(opt *pb.CreateVolumeOpts) (*model.VolumeSpec, erro
 		}
 	}
 
+        // Copy snapshot to volume
+        var snap = opt.GetSnapshotId()
```
Comment (wisererik) [inline contrib/drivers/lvm/lvm.go]: the default tab size should be 4.

**Comment by wisererik** [comment]
Hi @lixiaoy1  The API document should be updated which is located in [opensds/openapi-spec/swagger.yaml](https://github.com/opensds/opensds/tree/master/openapi-spec).

File: pkg/api/db.go [hunk 2]
```diff
@@ -44,6 +44,23 @@ func CreateVolumeDBEntry(ctx *c.Context, in *model.VolumeSpec) (*model.VolumeSpe
...
 		return nil, errors.New(errMsg)
 	}
+        if in.SnapshotId != "" {
+                snap, err := db.C.GetVolumeSnapshot(ctx, in.SnapshotId)
+                if err != nil {
+                        log.Error("Get snapshot failed in create volume method: ", err)
+                        return nil, err
+                }
+                if snap.Status != model.VolumeSnapAvailable {
+                        var errMsg = "Only the snapshot is available, the volume can be created"
```
Comment (xing-yang) [inline pkg/api/db.go]: s/Only the snapshot/Only if the snapshot

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -120,6 +150,8 @@ func (c *Controller) CreateVolume(ctx *c.Context, in *model.VolumeSpec, errchanV
 		AvailabilityZone: in.AvailabilityZone,
 		ProfileId:        profile.Id,
 		PoolId:           polInfo.Id,
+                SnapshotId:       in.SnapshotId,
+                SnapshotSize:     snapSize,
```
Comment (xing-yang) [inline pkg/controller/controller.go]: Is it necessary to pass "SnapshotSize" as a parameter to CreateVolumeOpts and compare that to volume size when creating the db entry?   Can you do the comparison here and if size is not right, don't proceed to create a db entry?

File: pkg/dock/proto/dock.proto [hunk 4]
```diff
@@ -112,6 +112,8 @@ message CreateVolumeOpts {
     map<string, string> replicationDriverData = 13;
     // The uuid of replication
     string replicationId = 14;
+    // The size of snapshot
+    int64 snapshotSize = 15;
```
Comment (xing-yang) [inline pkg/dock/proto/dock.proto]: See my previous comment.  It seems not necessary to add this snapshotSize in CreateVolumeOpts.  We can compare snapshot size with volume size earlier and bail out if it is not right.

File: contrib/drivers/lvm/lvm.go (hunk 1 above)
Comment (lixiaoy1) [reply contrib/drivers/lvm/lvm.go]: Done.

File: pkg/api/db.go (hunk 2 above)
Comment (lixiaoy1) [reply pkg/api/db.go]: Done

File: pkg/controller/controller.go (hunk 3 above)
Comment (lixiaoy1) [reply pkg/controller/controller.go]: We allow creating a volume from a snapshot whose size is bigger than the snapshot. So it needs the size of the snapshot in the driver, for example, lvm.go. Or else it needs to query the size of the snapshot in lvm.go. How do you think?

File: contrib/drivers/lvm/lvm.go [hunk 5]
```diff
@@ -110,6 +112,24 @@ func (d *Driver) CreateVolume(opt *pb.CreateVolumeOpts) (*model.VolumeSpec, erro
 		}
 	}
 
+	// Copy snapshot to volume
+	var snap = opt.GetSnapshotId()
```
Comment (leonwanghui) [inline contrib/drivers/lvm/lvm.go]: I think the operation of create volume and create volume from snapshot can not be executed at the same time, so you should add a switch flag to identify which operation should be performed.

Maybe I didn't understand, please explain what is `dd` command used for.

Comment (lixiaoy1) [reply contrib/drivers/lvm/lvm.go]: They will both be executed when creating from snapshots. lvcreate at first creates an lv, and then dd copies data from source snapshot lv to the new lv.  LVM has no command to create a thick lv from a snapshot directly.

Comment (leonwanghui) [reply contrib/drivers/lvm/lvm.go]: Got it, thanks!

File: pkg/controller/controller.go (hunk 3 above)
Comment (xing-yang) [reply pkg/controller/controller.go]: Got it.  Thanks!

**Pull request merged**